rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,499 @@
+
+###### Example: TypeScript 3.5 mitigation
+
+Assume an addon with the following type as its only export (which is quite weird as code goes, but captures the behavior for the example):
+
+```ts
+export default class FancyAddon {
+  fancy = true;
+}
+
+function lookup<T>(someValue: T) {
+  if (someValue == null) {
+    return new Promise((resolve) => {
+      resolve(someValue);
+    });
+  }
+  return Promise.resolve(new FancyAddon());
+}
+
+// THIS VALUE CHANGES BETWEEN 3.4 AND 3.5
+const foo = lookup("foo");
+```
+
+In TypeScript before 3.5, the result of `lookup("foo")` would be `Promise<{}>` ([playground][3.3-playground]). In TypeScript 3.5, the result would be `Promise<unknown>` ([playground][3.5-playground]). A user who had defined a type locally expecting `Promise<{}>` would now be forced to change her code:
+
+```ts
+interface ResultOfAddon {
+  foo: Promise<{}>;
+}
+
+let result: ResultOfAddon = {
+  foo: lookup("foo"), // TYPE ERROR IN 3.5
+};
+```
+
+[3.3-playground]: https://www.typescriptlang.org/play/?ts=3.3.3&ssl=1&ssc=1&pln=14&pc=27#code/KYDwDg9gTgLgBAE2AMwIYFcA28DGnUDOBcAYqgHY4CeAggghOXAN4CwAUHHGpVXALxwYUdMADcHAL4cOydJRgBLRnEwQIAa3RgAPABUAfAAoCEALbAAaqkyiAXHD0BKFhy6LkcE+as3RAwXIsTBc2Ti44KGAYdCgmcmAAdzgABShzRQJgIyMo00wAN2AXfgNXcIjI4Hyi7wtrW2KJCrhJJ2auaXComLjU9LNM4AA6PIhC7ITksl46BnIjJ3apGXYAejXHAAkASQBlOEsaABkAVQBROABhLZoAOQBxc4OAIXO9AHVz87u4AGZhgAWOD3AAi-2GAFYODhGAR4Mh1AJVOotGAjAAiREQDHtIA
+[3.5-playground]: https://www.typescriptlang.org/play/?ts=3.5.1&ssl=1&ssc=1&pln=14&pc=27#code/KYDwDg9gTgLgBAE2AMwIYFcA28DGnUDOBcAYqgHY4CeAggghOXAN4CwAUHHGpVXALxwYUdMADcHAL4cOydJRgBLRnEwQIAa3RgAPABUAfAAoCEALbAAaqkyiAXHD0BKFhy6LkcE+as3RAwXIsTBc2Ti44KGAYdCgmcmAAdzgABShzRQJgIyMo00wAN2AXfgNXcIjI4Hyi7wtrW2KJCrhJJ2auaXComLjU9LNM4AA6PIhC7ITksl46BnIjJ3apGXYAejXHAAkASQBlOEsaABkAVQBROABhLZoAOQBxc4OAIXO9AHVz87u4AGZhgAWOD3AAi-2GAFYODhGAR4Mh1AJVOotGAjAAiREQDHtIA
+
+If an addon author wanted to migitate this change, she would follow this procedure:
+
+1.  Backwards-compatibly *fix* the types by explicitly setting the return type on `lookup`:
+
+    ```diff
+    - function lookup<T>(someValue: T) {
+    + function lookup<T>(someValue: T): Promise<{}> {
+    ```
+
+2.  Create a new directory, named something like `ts3.5`.
+
+3.  Generate the type definition files for the addon by running `ember ts:precompile`.
+
+4.  Manually move the generated type definition files into `ts3.5`.
+
+5.  In the `ts3.5` directory, either *remove* or *change* the explicit return type, so that the default from TypeScript 3.5 is restored:
+
+    ```diff
+    - function lookup<T>(someValue: T): Promise<{}> {
+    + function lookup<T>(someValue: T): Promise<unknown> {
+    ```
+
+6.  Commit the `ts3.5` directory, since it now needs to be maintained manually until a breaking change of the library is released which opts into the new behavior.
```

Comment:
I think there would be a couple of additional steps required to ensure this works:
 - every file in `ts3.5` would need to be updated to be wrapped in a `declare module 'the-addon/the-actual-import-path' { ... }`
 - every file in `ts3.5` would need to be (directly or indirectly) imported from `ts3.5/index.d.ts` in order to be included for consideration

Reply:
Ahhh, yes – thank you for catching that. Will push a revision to that effect.

Reply:
Resolved in 5cfd8d4.